processing/src/main/java/io/druid/segment/StringDimensionMergerV9.java

```diff
@@ -0,0 +1,575 @@
+                           ? CompressedVSizeIndexedV3Writer.create(
+          ioPeon,
+          filenameBase,
+          cardinality,
+          compressionStrategy
+      )
+                           : new VSizeIndexedWriter(ioPeon, filenameBase, cardinality);
+    } else {
+      encodedValueWriter = (compressionStrategy != CompressedObjectStrategy.CompressionStrategy.UNCOMPRESSED)
+                           ? CompressedVSizeIntsIndexedWriter.create(
+          ioPeon,
+          filenameBase,
+          cardinality,
+          compressionStrategy
+      )
+                           : new VSizeIndexedIntsWriter(ioPeon, filenameBase, cardinality);
+    }
+    encodedValueWriter.open();
+  }
+
+
+  @Override
+  public int[] convertSegmentRowValuesToMergedRowValues(int[] segmentRow, int segmentIndexNumber)
+  {
+    int[] dimVals = (int[]) segmentRow;
+    // For strings, convert missing values to null/empty if conversion flag is set
+    // But if bitmap/dictionary is not used, always convert missing to 0
+    if (dimVals == null) {
+      return convertMissingValues ? EMPTY_STR_DIM_ARRAY : null;
+    }
+
+    int[] newDimVals = new int[dimVals.length];
+    IntBuffer converter = dimConversions.get(segmentIndexNumber);
+
+    for (int i = 0; i < dimVals.length; i++) {
+      if (converter != null) {
+        newDimVals[i] = converter.get(dimVals[i]);
+      } else {
+        newDimVals[i] = dimVals[i];
+      }
+    }
+
+    return newDimVals;
+  }
+
+  @Override
+  public void processMergedRow(int[] rowValues) throws IOException
+  {
+    int[] vals = (int[]) rowValues;
+    if (vals == null || vals.length == 0) {
+      nullRowsBitmap.add(rowCount);
+    } else if (hasNull && vals.length == 1 && (vals[0]) == 0) {
+      // Dictionary encoded, so it's safe to cast dim value to integer
+      // If this dimension has the null/empty str in its dictionary, a row with a single-valued dimension
+      // that matches the null/empty str's dictionary ID should also be added to nullRowBitmap.
+      nullRowsBitmap.add(rowCount);
+    }
+    processMergedRowHelper(vals);
+    rowCount++;
+  }
+
+  protected void processMergedRowHelper(int[] vals) throws IOException
+  {
+    encodedValueWriter.add(vals);
+  }
+
+  @Override
+  public void writeIndexes(List<IntBuffer> segmentRowNumConversions, Closer closer) throws IOException
+  {
+    long dimStartTime = System.currentTimeMillis();
+    final BitmapSerdeFactory bitmapSerdeFactory = indexSpec.getBitmapSerdeFactory();
+
+    String bmpFilename = String.format("%s.inverted", dimensionName);
+    bitmapWriter = new GenericIndexedWriter<>(
+        ioPeon,
+        bmpFilename,
+        indexSpec.getBitmapSerdeFactory().getObjectStrategy()
+    );
+    bitmapWriter.open();
+
+    // write dim values to one single file because we need to read it
+    File dimValueFile = IndexIO.makeDimFile(outDir, dimensionName);
+    FileOutputStream fos = new FileOutputStream(dimValueFile);
```

Comment:
these feel a little shaky without some sort of try block. Are these resources guaranteed to be cleaned up properly?